From 4a9f3bd09073fe62252eea6ab0385ad9e91f7bbb Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Mon, 8 Mar 2021 20:17:25 -0500
Subject: [PATCH] Disable OP_IFDUP (#65)

Rebase provenance: this historical change was adapted onto the pinned
Bitcoin ABC base. Source diff is the recorded Lotus V1 replay commit below.

Historical-source-commit: 597ab7cfd6a84dddaf8a66ad8cee43eeaf22e794
Recorded-replay-commit: 69b10982735c3db3baad09ff07a097e45ee5a6c0

diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index a6281e7fc1..fca99154c7 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -85,6 +85,7 @@ static bool IsOpcodeDisabled(opcodetype opcode, uint32_t flags) {
         case OP_VER:
         case OP_VERIF:
         case OP_VERNOTIF:
+        case OP_IFDUP:
         case OP_INVERT:
         case OP_RESERVED1:
         case OP_RESERVED2:
@@ -542,18 +543,6 @@ bool ScriptInterpreter::RunNextOpInner() {
                 swap(stacktop(-3), stacktop(-1));
             } break;
 
-            case OP_IFDUP: {
-                // (x - 0 | x x)
-                if (stack.size() < 1) {
-                    return set_error(serror,
-                                     ScriptError::INVALID_STACK_OPERATION);
-                }
-                valtype vch = stacktop(-1);
-                if (CastToBool(vch)) {
-                    stack.push_back(vch);
-                }
-            } break;
-
             case OP_DEPTH: {
                 // -- stacksize
                 CScriptNum bn(stack.size());
diff --git a/src/test/data/script_tests.json b/src/test/data/script_tests.json
index a52280c37a..a9aa27145c 100644
--- a/src/test/data/script_tests.json
+++ b/src/test/data/script_tests.json
@@ -81,9 +81,9 @@
 ["10 0 11", "TOALTSTACK DROP FROMALTSTACK ADD 21 EQUAL", "", "OK"],
 ["'gavin_was_here'", "TOALTSTACK 11 FROMALTSTACK 'gavin_was_here' EQUALVERIFY 11 EQUAL", "", "OK"],
 
-["0", "IFDUP DEPTH 1 EQUALVERIFY 0 EQUAL", "", "OK"],
-["1", "IFDUP DEPTH 2 EQUALVERIFY 1 EQUALVERIFY 1 EQUAL", "", "OK"],
-["0x05 0x0100000000", "IFDUP DEPTH 2 EQUALVERIFY 0x05 0x0100000000 EQUAL", "", "OK", "IFDUP dups non ints"],
+["0", "IFDUP DEPTH 1 EQUALVERIFY 0 EQUAL", "", "DISABLED_OPCODE"],
+["1", "IFDUP DEPTH 2 EQUALVERIFY 1 EQUALVERIFY 1 EQUAL", "", "DISABLED_OPCODE"],
+["0x05 0x0100000000", "IFDUP DEPTH 2 EQUALVERIFY 0x05 0x0100000000 EQUAL", "", "DISABLED_OPCODE", "IFDUP disabled"],
 ["0", "DROP DEPTH 0 EQUAL", "", "OK"],
 ["0", "DUP 1 ADD 1 EQUALVERIFY 0 EQUAL", "", "OK"],
 ["0 1", "NIP", "", "OK"],
@@ -387,7 +387,7 @@
 ["0 1 0 0", "2OVER", "", "OK"],
 ["0 1 0 0 0 0", "2ROT", "", "OK"],
 ["0 1 0 0", "2SWAP", "", "OK"],
-["1", "IFDUP", "", "OK"],
+["1", "IFDUP", "", "DISABLED_OPCODE"],
 ["", "DEPTH 1", "", "OK"],
 ["0", "DROP 1", "", "OK"],
 ["1", "DUP", "", "OK"],
@@ -819,7 +819,7 @@
 
 ["1 TOALTSTACK", "FROMALTSTACK 1", "", "SIG_PUSHONLY", "alt stack not shared between sig/pubkey; kept in case SIGPUSHONLY gets disabled again"],
 
-["", "IFDUP DEPTH 0 EQUAL", "", "INVALID_STACK_OPERATION"],
+["", "IFDUP DEPTH 0 EQUAL", "", "DISABLED_OPCODE"],
 ["", "DROP DEPTH 0 EQUAL", "", "INVALID_STACK_OPERATION"],
 ["", "DUP DEPTH 0 EQUAL", "", "INVALID_STACK_OPERATION"],
 ["1", "DUP 1 ADD 2 EQUALVERIFY 0 EQUAL", "", "EVAL_FALSE"],
@@ -1734,7 +1734,7 @@
 ["1 1 1", "2OVER", "", "INVALID_STACK_OPERATION"],
 ["1 1 1 1 1", "2ROT", "", "INVALID_STACK_OPERATION"],
 ["1 1 1", "2SWAP", "", "INVALID_STACK_OPERATION"],
-["", "IFDUP 1", "", "INVALID_STACK_OPERATION"],
+["", "IFDUP 1", "", "DISABLED_OPCODE"],
 ["", "DROP 1", "", "INVALID_STACK_OPERATION"],
 ["", "DUP 1", "", "INVALID_STACK_OPERATION"],
 ["1", "NIP", "", "INVALID_STACK_OPERATION"],
